### PR TITLE
Make CharacterCategory a bitset

### LINKED ISSUE #54: Make CharacterCategory a bitset
Part of #36 
Related to #53 

Rightnow it is a hashset of enums which is super inefficient compared to bitsets.

Making it a bitset will allow to make computational complexity of many algorithms to O(1) from O(n) as with hashmaps and remove allocations

### PR DESCRIPTION
**Pull request by eiennohito**
fixes #54

also cleaned up tests that were failing

### CODE REVIEW

File: sudachi/src/dic/character_category.rs [hunk 1]
```diff
@@ -46,25 +42,23 @@ pub enum Error {
...
+    categories: CategoryType,
 }
 
 /// CharacterCategory holds mapping from character to character category type
-#[derive(Debug, Default, Clone)]
+#[derive(Debug, Clone)]
 pub struct CharacterCategory {
-    ranges: Vec<Range>,
+    boundaries: Vec<u32>,
+    categories: Vec<CategoryType>,
```
Comment (mh-northlander) [inline sudachi/src/dic/character_category.rs:63]: I feel it's better to have a comment about how boundaries and categories correspond to.
(we have `boundaries.len()+1` categories and `categories[i]` is for range `boundaries[i-1]..boundaries[i]`, right?)

File: sudachi/src/dic/category_type.rs [hunk 2]
```diff
@@ -14,76 +14,154 @@
...
+use crate::error::SudachiError;
+use bitflags::bitflags;
 use std::str::FromStr;
 
-use crate::error::SudachiError;
+bitflags! {
+    /// A set of categories for a character
+    ///
+    /// Implemented as a bitset with fixed size
+    #[repr(transparent)]
```
Comment (mh-northlander) [inline sudachi/src/dic/category_type.rs:25]: What is this attribute for? (I'm just curious.)

File: sudachi/src/input_text/utf8_input_text_builder.rs [hunk 3]
```diff
@@ -152,13 +152,11 @@ impl Utf8InputTextBuilder<'_, '_> {
 
             // in rust, char corresponds to unicode scalar value
```
Comment (mh-northlander) [inline sudachi/src/input_text/utf8_input_text_builder.rs:153]: I added this comment mentioning to java implementation.
Maybe it's ok to remove it.

**Review by mh-northlander — APPROVED** [review]

File: sudachi/src/input_text/utf8_input_text_builder.rs (hunk 3 above)
Comment (eiennohito) [reply sudachi/src/input_text/utf8_input_text_builder.rs:153]: OK

File: sudachi/src/dic/category_type.rs (hunk 2 above)
Comment (eiennohito) [reply sudachi/src/dic/category_type.rs:25]: From the binary point of view treat it as if it was a single u32, so this struct and related things actually do not exist in the binary at all, it's only single u32 which is usually passed via registers and does not even go to the stack most of the time.

Comment (eiennohito) [reply sudachi/src/dic/category_type.rs:25]: In Rust this attribute can be assigned to any struct which has exactly one field.
https://doc.rust-lang.org/nomicon/other-reprs.html#reprtransparent

File: sudachi/src/dic/character_category.rs (hunk 1 above)
Comment (eiennohito) [reply sudachi/src/dic/character_category.rs:63]: Certainly, will add.

File: sudachi/src/dic/category_type.rs (hunk 2 above)
Comment (mh-northlander) [reply sudachi/src/dic/category_type.rs:25]: I got it, thank you

**Pull request merged**